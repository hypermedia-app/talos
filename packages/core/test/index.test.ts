import path from 'path'
import url from 'url'
import chai, { expect } from 'chai'
import formats from '@rdfjs-elements/formats-pretty'
import type { Dataset } from '@zazuko/env/lib/DatasetExt.js'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import $rdf from '../env.js'
import { fromDirectories } from '../index.js'

const testDir = url.fileURLToPath(new URL('../../../test-resources', import.meta.url))
const ns = $rdf.namespace('https://example.com')

$rdf.formats.import({
  serializers: formats.serializers,
})

describe('@hydrofoil/talos-core', () => {
  chai.use(jestSnapshotPlugin())

  describe('fromDirectories', () => {
    let dataset: Dataset

    beforeEach(async () => {
      const dirs = [
        path.resolve(testDir, './resources'),
        path.resolve(testDir, './resources.foo'),
        path.resolve(testDir, './resources.bar'),
      ]
      dataset = await fromDirectories(dirs, ns().value)
    })

    it('merges resources from multiple graph documents', async function () {
      const resource = dataset.match(null, null, null, ns())

      expect(await resource.serialize({ format: 'application/trig' })).toMatchSnapshot()
    })

    it('merges resources from dataset and graph documents', async function () {
      const resource = dataset.match(null, null, null, ns('/trig/users/john-doe'))

      expect(await resource.serialize({ format: 'application/trig' })).toMatchSnapshot()
    })

    it('merges resources from multiple dataset documents', async function () {
      const resource = dataset.match(null, null, null, ns('/trig/users/jane-doe'))

      expect(await resource.serialize({ format: 'application/trig' })).toMatchSnapshot()
    })

    it('marks a resource for "overwrite" by default', () => {
      const match = dataset.match(ns(), $rdf.ns.talos.action, null, $rdf.ns.talos.resources)
      const [{ object: action }, ...more] = match

      expect(action).to.deep.equal($rdf.ns.talos.overwrite)
      expect(more).to.be.empty
    })

    it('marks a resource for "merge" when prefix is used', () => {
      const datasetCore = dataset.match(ns('/project/creta/user.group/admins'), $rdf.ns.talos.action, null, $rdf.ns.talos.resources)
      const [{ object: action }, ...more] = datasetCore

      expect(action.equals($rdf.ns.talos.merge)).to.be.true
      expect(more).to.be.empty
    })

    it('uses the last representation when is marked to replace other envs', async function () {
      const resource = dataset.match(null, null, null, ns('/only/one'))

      expect(await resource.serialize({ format: 'application/trig' })).toMatchSnapshot()
    })

    it('uses the last representation when is marked in another env', async function () {
      const resource = dataset.match(null, null, null, ns('/only/two'))

      expect(await resource.serialize({ format: 'application/trig' })).toMatchSnapshot()
    })

    it('uses the last representation when is marked in a dataset document', async function () {
      const resource = dataset.match(null, null, null, ns('/only/three'))

      expect(await resource.serialize({ format: 'application/trig' })).toMatchSnapshot()
    })

    it('adds data with SPARQL', async function () {
      const resource = dataset.match(null, null, null, ns('/trig/users/john-doe-additional'))

      expect(await resource.serialize({ format: 'application/trig' })).toMatchSnapshot()
    })

    it('successfully executed federated query', () => {
      const cubes = $rdf.clownface({
        dataset: dataset.match(null, null, null, ns('/lindas-cubes')),
      }).has($rdf.ns.rdf.type, $rdf.namedNode('https://cube.link/Cube'))

      expect(cubes.values).to.have.property('length').greaterThan(0)
    })

    it('marks generated resources to overwrite', () => {
      const [action] = dataset.match(ns('/lindas-cubes'), $rdf.ns.talos.action, null, $rdf.ns.talos.resources)

      expect(action.object).to.deep.equal($rdf.ns.talos.overwrite)
    })
  })
})
