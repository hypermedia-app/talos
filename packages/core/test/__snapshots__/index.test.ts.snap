// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@hydrofoil/talos-core fromDirectories adds data with SPARQL 1`] = `
"
<https://example.com/trig/users/john-doe-additional> {
	<https://example.com/trig/users/john-doe> a <http://xmlns.com/foaf/0.1/Person> ;
		<http://schema.org/givenName> \\"John\\" .
}

"
`;

exports[`@hydrofoil/talos-core fromDirectories merges resources from dataset and graph documents 1`] = `
"
<https://example.com/trig/users/john-doe> {
	<https://example.com/trig/users/john-doe> a <http://schema.org/Person>, <http://xmlns.com/foaf/0.1/Person> ;
		<http://www.w3.org/ns/earl#test> \\"trig\\" ;
		<http://xmlns.com/foaf/0.1/name> \\"John Doe\\" ;
		<http://schema.org/givenName> \\"John\\" .
}

"
`;

exports[`@hydrofoil/talos-core fromDirectories merges resources from multiple dataset documents 1`] = `
"
<https://example.com/trig/users/jane-doe> {
	<https://example.com/trig/users/jane-doe> a <http://schema.org/Person>, <http://xmlns.com/foaf/0.1/Person> ;
		<http://www.w3.org/ns/earl#test> \\"trig\\" ;
		<http://xmlns.com/foaf/0.1/name> \\"Jane Doe\\" .
}

"
`;

exports[`@hydrofoil/talos-core fromDirectories merges resources from multiple graph documents 1`] = `
"
<https://example.com> {
	<https://example.com> a <http://schema.org/Thing> ;
		<http://schema.org/hasPart> <https://example.com/#part>, _:b1, _:namedPart, _:b6 ;
		<http://schema.org/name> \\"Bar environment\\" .

	<https://example.com/#part> <http://schema.org/minValue> \\"10\\"^^<http://www.w3.org/2001/XMLSchema#integer> ;
		<http://schema.org/maxValue> \\"100\\"^^<http://www.w3.org/2001/XMLSchema#integer> .

	_:b1 <http://schema.org/name> \\"totally anonymous\\" .

	_:namedPart <http://schema.org/name> \\"Foo part\\", \\"Bar part\\" .

	_:b6 <http://schema.org/name> \\"totally anonymous\\" .
}

"
`;

exports[`@hydrofoil/talos-core fromDirectories uses the last representation when is marked in a dataset document 1`] = `
"
<https://example.com/only/three> {
	<https://example.com/only/three> <http://schema.org/name> \\"bar\\" .
}

"
`;

exports[`@hydrofoil/talos-core fromDirectories uses the last representation when is marked in another env 1`] = `
"
<https://example.com/only/two> {
	<https://example.com/only/two> <http://schema.org/name> \\"bar\\" .
}

"
`;

exports[`@hydrofoil/talos-core fromDirectories uses the last representation when is marked to replace other envs 1`] = `
"
<https://example.com/only/one> {
	<https://example.com/only/one> <http://schema.org/name> \\"bar\\" .
}

"
`;
